velox/exec/HashTable.cpp

```diff
@@ -1682,42 +1695,62 @@ int32_t HashTable<ignoreNullKeys>::listJoinResults(
   int numOut = 0;
   auto maxOut = inputRows.size();
   while (iter.lastRowIndex < iter.rows->size()) {
-    if (!iter.nextHit) {
-      auto row = (*iter.rows)[iter.lastRowIndex];
-      iter.nextHit = (*iter.hits)[row]; // NOLINT
-      if (!iter.nextHit) {
-        ++iter.lastRowIndex;
-
-        if (includeMisses) {
-          inputRows[numOut] = row; // NOLINT
-          hits[numOut] = nullptr;
-          ++numOut;
-          if (numOut >= maxOut) {
-            return numOut;
-          }
+    auto row = (*iter.rows)[iter.lastRowIndex];
+    auto hit = (*iter.hits)[row]; // NOLINT
+    if (!hit) {
+      ++iter.lastRowIndex;
+      if (includeMisses) {
+        inputRows[numOut] = row; // NOLINT
+        hits[numOut] = nullptr;
+        ++numOut;
+        if (numOut >= maxOut) {
+          return numOut;
         }
-        continue;
       }
+      continue;
     }
-    while (iter.nextHit) {
-      char* next = nullptr;
-      if (nextOffset_) {
-        next = nextRow(iter.nextHit);
-        if (next) {
-          __builtin_prefetch(reinterpret_cast<char*>(next) + nextOffset_);
+
+    auto numRows = -1;
+    auto rows = rows_->getNextRowVector(hit);
+    if (!rows) {
+      inputRows[numOut] = row; // NOLINT
+      hits[numOut] = hit;
+      numOut++;
+      iter.lastRowIndex++;
+    } else {
+      numRows = rows->size();
+      constexpr int32_t kWidth = xsimd::batch<int64_t>::size;
+      constexpr int32_t kIntWidth = kWidth * 2;
+      if (numRows >= kIntWidth) {
+        auto sourceHits = reinterpret_cast<int64_t*>(rows->data());
+        auto resultHits = reinterpret_cast<int64_t*>(hits.data());
+        auto indices = reinterpret_cast<int32_t*>(inputRows.data());
+        auto indexVector = xsimd::batch<int32_t>::broadcast(row);
+        for (; iter.lastDuplicateRowIndex + kIntWidth <= numRows &&
+             numOut + kIntWidth <= maxOut;
+             iter.lastDuplicateRowIndex += kIntWidth, numOut += kIntWidth) {
+          indexVector.store_unaligned(indices + numOut);
```

Comment:
How much worse if we just use `std::fill` and `memcpy` here?  Can you measure it with the benchmark?

Reply:
There is no much difference, fill+ memcpy is little slower, BTW std::fill_n is faster than std::fill in my test.
This is result for before simd solution:
```
============================================================================
[...]marks/HashJoinListResultBenchmark.cpp     relative  time/iter   iters/s
============================================================================
array_1_20%:1;80%:0;                                         2.09s   477.90m
array_1_20%:5;80%:0;                                         2.98s   336.10m
array_1_20%:10;80%:0;                                        3.31s   301.82m
array_1_20%:20;80%:0;                                        4.28s   233.49m
array_1_20%:50;80%:0;                                        6.50s   153.87m
array_1_10%:5;10%:1;80%:0;                                   2.89s   346.41m
array_1_10%:10;10%:5;10%:1;70%:0;                            3.78s   264.49m
array_1_10%:20;10%:10;10%:5;10%:1;60%:0;                     5.66s   176.69m
array_1_10%:50;10%:20;10%:10;10%:5;10%:1;50%:0;              7.93s   126.05m
normalized key_1_20%:1;80%:0;                                3.96s   252.64m
normalized key_1_20%:5;80%:0;                                5.27s   189.60m
normalized key_1_20%:10;80%:0;                               5.57s   179.68m
normalized key_1_20%:20;80%:0;                               6.99s   143.03m
normalized key_1_20%:50;80%:0;                               9.28s   107.74m
normalized key_1_10%:5;10%:1;80%:0;                          4.68s   213.81m
normalized key_1_10%:10;10%:5;10%:1;70%:0;                   5.53s   180.67m
normalized key_1_10%:20;10%:10;10%:5;10%:1;60%:              7.10s   140.85m
normalized key_1_10%:50;10%:20;10%:10;10%:5;10%              9.62s   103.92m
hash_1_20%:1;80%:0;                                          5.48s   182.37m
hash_1_20%:5;80%:0;                                          6.53s   153.24m
hash_1_20%:10;80%:0;                                         6.84s   146.12m
hash_1_20%:20;80%:0;                                         8.06s   124.09m
hash_1_20%:50;80%:0;                                        10.99s    90.99m
hash_1_10%:5;10%:1;80%:0;                                    6.69s   149.54m
hash_1_10%:10;10%:5;10%:1;70%:0;                             7.39s   135.29m
hash_1_10%:20;10%:10;10%:5;10%:1;60%:0;                      9.00s   111.11m
hash_1_10%:50;10%:20;10%:10;10%:5;10%:1;50%:0;              11.87s    84.24m
*** Results:
HashTableSize:2097149, BuildInputSize:2516577, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 20%:1; 80%:0;
 mode=ARRAY numOutput=118279166 totalClock=6.2674e+09 listJoinResultClocks=3.83657e+09(61.2147%) buildClocks=1.91906e+08(3.06198%)
HashTableSize:2097149, BuildInputSize:4194293, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 20%:5; 80%:0;
 mode=ARRAY numOutput=197131818 totalClock=8.91163e+09 listJoinResultClocks=5.66116e+09(63.5256%) buildClocks=8.64172e+08(9.69713%)
HashTableSize:2097149, BuildInputSize:6291438, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 20%:10; 80%:0;
 mode=ARRAY numOutput=295697633 totalClock=9.92378e+09 listJoinResultClocks=6.03553e+09(60.8189%) buildClocks=1.25143e+09(12.6104%)
HashTableSize:2097149, BuildInputSize:10485728, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 20%:20; 80%:0;
 mode=ARRAY numOutput=492829263 totalClock=1.28277e+10 listJoinResultClocks=7.6725e+09(59.812%) buildClocks=1.95012e+09(15.2024%)
HashTableSize:2097149, BuildInputSize:23068598, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 20%:50; 80%:0;
 mode=ARRAY numOutput=1084224153 totalClock=1.9466e+10 listJoinResultClocks=1.04702e+10(53.7874%) buildClocks=3.94053e+09(20.2432%)
HashTableSize:2097149, BuildInputSize:3355431, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 10%:5; 10%:1; 80%:0;
 mode=ARRAY numOutput=157705398 totalClock=8.64646e+09 listJoinResultClocks=5.51262e+09(63.7558%) buildClocks=7.23396e+08(8.36639%)
HashTableSize:2097149, BuildInputSize:5452570, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 10%:10; 10%:5; 10%:1; 70%:0;
 mode=ARRAY numOutput=256271213 totalClock=1.13243e+10 listJoinResultClocks=7.25108e+09(64.0313%) buildClocks=1.36248e+09(12.0315%)
HashTableSize:2097149, BuildInputSize:9646849, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 10%:20; 10%:10; 10%:5; 10%:1; 60%:0;
 mode=ARRAY numOutput=453402843 totalClock=1.69518e+10 listJoinResultClocks=1.07462e+10(63.3928%) buildClocks=2.83384e+09(16.7171%)
HashTableSize:2097149, BuildInputSize:20132548, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 10%:50; 10%:20; 10%:10; 10%:5; 10%:1; 50%:0;
 mode=ARRAY numOutput=946231683 totalClock=2.37624e+10 listJoinResultClocks=1.44833e+10(60.9507%) buildClocks=4.46488e+09(18.7897%)
HashTableSize:2097149, BuildInputSize:2516577, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 20%:1; 80%:0;
 mode=NORMALIZED_KEY numOutput=118279119 totalClock=1.18556e+10 listJoinResultClocks=4.15686e+09(35.0624%) buildClocks=2.09615e+08(1.76806%)
HashTableSize:2097149, BuildInputSize:4194293, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 20%:5; 80%:0;
 mode=NORMALIZED_KEY numOutput=197131771 totalClock=1.57976e+10 listJoinResultClocks=6.21815e+09(39.3614%) buildClocks=5.12045e+08(3.24128%)
HashTableSize:2097149, BuildInputSize:6291438, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 20%:10; 80%:0;
 mode=NORMALIZED_KEY numOutput=295697586 totalClock=1.66698e+10 listJoinResultClocks=6.75414e+09(40.5171%) buildClocks=6.6058e+08(3.96273%)
HashTableSize:2097149, BuildInputSize:10485728, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 20%:20; 80%:0;
 mode=NORMALIZED_KEY numOutput=492829216 totalClock=2.09404e+10 listJoinResultClocks=8.70678e+09(41.5788%) buildClocks=1.47948e+09(7.0652%)
HashTableSize:2097149, BuildInputSize:23068598, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 20%:50; 80%:0;
 mode=NORMALIZED_KEY numOutput=1084224106 totalClock=2.77998e+10 listJoinResultClocks=1.1297e+10(40.6369%) buildClocks=3.53038e+09(12.6993%)
HashTableSize:2097149, BuildInputSize:3355431, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 10%:5; 10%:1; 80%:0;
 mode=NORMALIZED_KEY numOutput=157705351 totalClock=1.40089e+10 listJoinResultClocks=5.55854e+09(39.6787%) buildClocks=3.25877e+08(2.32622%)
HashTableSize:2097149, BuildInputSize:5452570, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 10%:10; 10%:5; 10%:1; 70%:0;
 mode=NORMALIZED_KEY numOutput=256271166 totalClock=1.65781e+10 listJoinResultClocks=7.08177e+09(42.7177%) buildClocks=7.50849e+08(4.52917%)
HashTableSize:2097149, BuildInputSize:9646849, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 10%:20; 10%:10; 10%:5; 10%:1; 60%:0;
 mode=NORMALIZED_KEY numOutput=453402796 totalClock=2.12653e+10 listJoinResultClocks=9.97541e+09(46.9094%) buildClocks=1.15651e+09(5.43851%)
HashTableSize:2097149, BuildInputSize:20132548, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 10%:50; 10%:20; 10%:10; 10%:5; 10%:1; 50%:0;
 mode=NORMALIZED_KEY numOutput=946231636 totalClock=2.88219e+10 listJoinResultClocks=1.41126e+10(48.9648%) buildClocks=2.58081e+09(8.95436%)
HashTableSize:2097149, BuildInputSize:2516577, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 20%:1; 80%:0;
 mode=HASH numOutput=118279119 totalClock=1.64241e+10 listJoinResultClocks=3.9049e+09(23.7754%) buildClocks=1.78391e+08(1.08615%)
HashTableSize:2097149, BuildInputSize:4194293, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 20%:5; 80%:0;
 mode=HASH numOutput=197131771 totalClock=1.95456e+10 listJoinResultClocks=5.75096e+09(29.4232%) buildClocks=5.06579e+08(2.59177%)
HashTableSize:2097149, BuildInputSize:6291438, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 20%:10; 80%:0;
 mode=HASH numOutput=295697586 totalClock=2.04976e+10 listJoinResultClocks=6.25474e+09(30.5145%) buildClocks=8.0947e+08(3.9491%)
HashTableSize:2097149, BuildInputSize:10485728, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 20%:20; 80%:0;
 mode=HASH numOutput=492829216 totalClock=2.41369e+10 listJoinResultClocks=8.14702e+09(33.7534%) buildClocks=1.3305e+09(5.51229%)
HashTableSize:2097149, BuildInputSize:23068598, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 20%:50; 80%:0;
 mode=HASH numOutput=1084224106 totalClock=3.29194e+10 listJoinResultClocks=1.15513e+10(35.0895%) buildClocks=3.64667e+09(11.0776%)
HashTableSize:2097149, BuildInputSize:3355431, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 10%:5; 10%:1; 80%:0;
 mode=HASH numOutput=157705351 totalClock=2.0029e+10 listJoinResultClocks=5.62672e+09(28.0929%) buildClocks=3.57486e+08(1.78484%)
HashTableSize:2097149, BuildInputSize:5452570, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 10%:10; 10%:5; 10%:1; 70%:0;
 mode=HASH numOutput=256271166 totalClock=2.21386e+10 listJoinResultClocks=7.52525e+09(33.9915%) buildClocks=6.1594e+08(2.7822%)
HashTableSize:2097149, BuildInputSize:9646849, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 10%:20; 10%:10; 10%:5; 10%:1; 60%:0;

 mode=HASH numOutput=453402796 totalClock=2.69572e+10 listJoinResultClocks=1.04387e+10(38.7232%) buildClocks=1.30895e+09(4.85567%)
HashTableSize:2097149, BuildInputSize:20132548, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 10%:50; 10%:20; 10%:10; 10%:5; 10%:1; 50%:0;
 mode=HASH numOutput=946231636 totalClock=3.55568e+10 listJoinResultClocks=1.51703e+10(42.665%) buildClocks=3.18248e+09(8.9504%)


```

This is result of fill + memcpy:
```
============================================================================
[...]marks/HashJoinListResultBenchmark.cpp     relative  time/iter   iters/s
============================================================================
array_1_20%:1;80%:0;                                         2.25s   443.59m
array_1_20%:5;80%:0;                                         3.05s   327.57m
array_1_20%:10;80%:0;                                        3.46s   289.18m
array_1_20%:20;80%:0;                                        4.45s   224.54m
array_1_20%:50;80%:0;                                        6.33s   157.93m
array_1_10%:5;10%:1;80%:0;                                   2.73s   366.02m
array_1_10%:10;10%:5;10%:1;70%:0;                            3.99s   250.34m
array_1_10%:20;10%:10;10%:5;10%:1;60%:0;                     5.99s   167.08m
array_1_10%:50;10%:20;10%:10;10%:5;10%:1;50%:0;              8.78s   113.92m
normalized key_1_20%:1;80%:0;                                4.08s   244.98m
normalized key_1_20%:5;80%:0;                                4.91s   203.70m
normalized key_1_20%:10;80%:0;                               5.56s   180.00m
normalized key_1_20%:20;80%:0;                               6.92s   144.59m
normalized key_1_20%:50;80%:0;                               8.88s   112.60m
normalized key_1_10%:5;10%:1;80%:0;                          4.31s   231.98m
normalized key_1_10%:10;10%:5;10%:1;70%:0;                   5.76s   173.73m
normalized key_1_10%:20;10%:10;10%:5;10%:1;60%:              7.50s   133.42m
normalized key_1_10%:50;10%:20;10%:10;10%:5;10%             10.08s    99.19m
hash_1_20%:1;80%:0;                                          5.48s   182.46m
hash_1_20%:5;80%:0;                                          6.58s   151.92m
hash_1_20%:10;80%:0;                                         7.50s   133.39m
hash_1_20%:20;80%:0;                                         7.79s   128.35m
hash_1_20%:50;80%:0;                                         9.87s   101.28m
hash_1_10%:5;10%:1;80%:0;                                    5.99s   166.88m
hash_1_10%:10;10%:5;10%:1;70%:0;                             7.15s   139.86m
hash_1_10%:20;10%:10;10%:5;10%:1;60%:0;                      9.10s   109.88m
hash_1_10%:50;10%:20;10%:10;10%:5;10%:1;50%:0;              11.43s    87.52m
*** Results:
HashTableSize:2097149, BuildInputSize:2516577, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 20%:1; 80%:0;
 mode=ARRAY numOutput=118279166 totalClock=6.75219e+09 listJoinResultClocks=4.44252e+09(65.7937%) buildClocks=1.79594e+08(2.65979%)
HashTableSize:2097149, BuildInputSize:4194293, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 20%:5; 80%:0;
 mode=ARRAY numOutput=197131818 totalClock=9.14362e+09 listJoinResultClocks=6.13532e+09(67.0995%) buildClocks=6.91243e+08(7.55984%)
HashTableSize:2097149, BuildInputSize:6291438, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 20%:10; 80%:0;
 mode=ARRAY numOutput=295697633 totalClock=1.03576e+10 listJoinResultClocks=6.78596e+09(65.5167%) buildClocks=1.02219e+09(9.86898%)
HashTableSize:2097149, BuildInputSize:10485728, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 20%:20; 80%:0;
 mode=ARRAY numOutput=492829263 totalClock=1.3339e+10 listJoinResultClocks=8.334e+09(62.4786%) buildClocks=1.95623e+09(14.6656%)
HashTableSize:2097149, BuildInputSize:23068598, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 20%:50; 80%:0;
 mode=ARRAY numOutput=1084224153 totalClock=1.8965e+10 listJoinResultClocks=1.08686e+10(57.3087%) buildClocks=3.38753e+09(17.862%)
HashTableSize:2097149, BuildInputSize:3355431, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 10%:5; 10%:1; 80%:0;
 mode=ARRAY numOutput=157705398 totalClock=8.18319e+09 listJoinResultClocks=5.48802e+09(67.0645%) buildClocks=4.27986e+08(5.23006%)
HashTableSize:2097149, BuildInputSize:5452570, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 10%:10; 10%:5; 10%:1; 70%:0;
 mode=ARRAY numOutput=256271213 totalClock=1.19646e+10 listJoinResultClocks=8.29141e+09(69.2995%) buildClocks=1.08837e+09(9.0966%)
HashTableSize:2097149, BuildInputSize:9646849, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 10%:20; 10%:10; 10%:5; 10%:1; 60%:0;
 mode=ARRAY numOutput=453402843 totalClock=1.79268e+10 listJoinResultClocks=1.16284e+10(64.866%) buildClocks=2.62209e+09(14.6267%)
HashTableSize:2097149, BuildInputSize:20132548, ProbeInputSize:100000000, ExpectHashMode:ARRAY, KeyRepeatTimesDistribution: 10%:50; 10%:20; 10%:10; 10%:5; 10%:1; 50%:0;
 mode=ARRAY numOutput=946231683 totalClock=2.62921e+10 listJoinResultClocks=1.67434e+10(63.6822%) buildClocks=5.18348e+09(19.715%)
HashTableSize:2097149, BuildInputSize:2516577, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 20%:1; 80%:0;
 mode=NORMALIZED_KEY numOutput=118279119 totalClock=1.22265e+10 listJoinResultClocks=4.87373e+09(39.8622%) buildClocks=1.97598e+08(1.61615%)
HashTableSize:2097149, BuildInputSize:4194293, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 20%:5; 80%:0;
 mode=NORMALIZED_KEY numOutput=197131771 totalClock=1.47043e+10 listJoinResultClocks=6.23396e+09(42.3956%) buildClocks=5.00366e+08(3.40286%)
HashTableSize:2097149, BuildInputSize:6291438, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 20%:10; 80%:0;
 mode=NORMALIZED_KEY numOutput=295697586 totalClock=1.66395e+10 listJoinResultClocks=7.45026e+09(44.7744%) buildClocks=6.32316e+08(3.80008%)
HashTableSize:2097149, BuildInputSize:10485728, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 20%:20; 80%:0;
 mode=NORMALIZED_KEY numOutput=492829216 totalClock=2.07158e+10 listJoinResultClocks=9.35212e+09(45.1449%) buildClocks=1.15001e+09(5.55139%)
HashTableSize:2097149, BuildInputSize:23068598, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 20%:50; 80%:0;
 mode=NORMALIZED_KEY numOutput=1084224106 totalClock=2.65998e+10 listJoinResultClocks=1.14885e+10(43.1903%) buildClocks=3.22792e+09(12.1352%)
HashTableSize:2097149, BuildInputSize:3355431, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 10%:5; 10%:1; 80%:0;
 mode=NORMALIZED_KEY numOutput=157705351 totalClock=1.29113e+10 listJoinResultClocks=5.70044e+09(44.1509%) buildClocks=2.6661e+08(2.06494%)
HashTableSize:2097149, BuildInputSize:5452570, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 10%:10; 10%:5; 10%:1; 70%:0;
 mode=NORMALIZED_KEY numOutput=256271166 totalClock=1.72407e+10 listJoinResultClocks=8.50985e+09(49.359%) buildClocks=5.66882e+08(3.28804%)
HashTableSize:2097149, BuildInputSize:9646849, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 10%:20; 10%:10; 10%:5; 10%:1; 60%:0;
 mode=NORMALIZED_KEY numOutput=453402796 totalClock=2.24502e+10 listJoinResultClocks=1.13614e+10(50.6074%) buildClocks=1.46317e+09(6.51743%)
HashTableSize:2097149, BuildInputSize:20132548, ProbeInputSize:100000000, ExpectHashMode:NORMALIZED_KEY, KeyRepeatTimesDistribution: 10%:50; 10%:20; 10%:10; 10%:5; 10%:1; 50%:0;
 mode=NORMALIZED_KEY numOutput=946231636 totalClock=3.01968e+10 listJoinResultClocks=1.61713e+10(53.553%) buildClocks=2.47149e+09(8.18462%)
HashTableSize:2097149, BuildInputSize:2516577, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 20%:1; 80%:0;
 mode=HASH numOutput=118279119 totalClock=1.64154e+10 listJoinResultClocks=4.81181e+09(29.3127%) buildClocks=1.80661e+08(1.10056%)
HashTableSize:2097149, BuildInputSize:4194293, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 20%:5; 80%:0;
 mode=HASH numOutput=197131771 totalClock=1.97157e+10 listJoinResultClocks=6.41485e+09(32.5367%) buildClocks=4.65014e+08(2.35859%)
HashTableSize:2097149, BuildInputSize:6291438, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 20%:10; 80%:0;
 mode=HASH numOutput=295697586 totalClock=2.24552e+10 listJoinResultClocks=7.94121e+09(35.3647%) buildClocks=7.42141e+08(3.30499%)
HashTableSize:2097149, BuildInputSize:10485728, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 20%:20; 80%:0;
 mode=HASH numOutput=492829216 totalClock=2.33364e+10 listJoinResultClocks=8.35615e+09(35.8074%) buildClocks=1.46253e+09(6.26717%)
HashTableSize:2097149, BuildInputSize:23068598, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 20%:50; 80%:0;
 mode=HASH numOutput=1084224106 totalClock=2.95735e+10 listJoinResultClocks=1.09066e+10(36.8797%) buildClocks=3.0945e+09(10.4638%)
HashTableSize:2097149, BuildInputSize:3355431, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 10%:5; 10%:1; 80%:0;
 mode=HASH numOutput=157705351 totalClock=1.79484e+10 listJoinResultClocks=5.95425e+09(33.1743%) buildClocks=3.35566e+08(1.86962%)
HashTableSize:2097149, BuildInputSize:5452570, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 10%:10; 10%:5; 10%:1; 70%:0;
 mode=HASH numOutput=256271166 totalClock=2.14165e+10 listJoinResultClocks=8.26284e+09(38.5818%) buildClocks=5.84678e+08(2.73004%)
HashTableSize:2097149, BuildInputSize:9646849, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 10%:20; 10%:10; 10%:5; 10%:1; 60%:0;

 mode=HASH numOutput=453402796 totalClock=2.72581e+10 listJoinResultClocks=1.19411e+10(43.8075%) buildClocks=1.357e+09(4.97836%)
HashTableSize:2097149, BuildInputSize:20132548, ProbeInputSize:100000000, ExpectHashMode:HASH, KeyRepeatTimesDistribution: 10%:50; 10%:20; 10%:10; 10%:5; 10%:1; 50%:0;
 mode=HASH numOutput=946231636 totalClock=3.42237e+10 listJoinResultClocks=1.56527e+10(45.7363%) buildClocks=2.70059e+09(7.89097%)
```